File: PHP/Builder/CarBuilder.php
<?php
namespace DesignPatterns\Builder;
class CarBuilder implements VehicleBuilder
{
    /**
     * @var Vehicle
     */
    private $car;
    public function __construct()
    {
        $this->car = new Vehicle();
    }
   
    public function setEngine(string $enginType)
    {
        $this->car->setEngine($enginType);
    }
    
    public function  setWheels(int $wheels)
    {
        $this->car->setWheels($wheels);
    }

    public function setColor(string $color)
    {
        $this->car->setColor($color);
    }

    public function getVehicle(): Vehicle
    {
        return $this->car;
    }
    
    
}

File: Go/builder/CarBuilder.go
package builder
type CarBuilder struct {
	v Vehicle
}

func (c *CarBuilder) SetWheels() VehicleBuilder {
	c.v.Wheels = 4
	return c
}

func (c *CarBuilder) SetSeats() VehicleBuilder {
	c.v.Seats = 5
	return c
}

func (b *CarBuilder) TopSpeed() VehicleBuilder {
	b.v.TopSpeed = 160
	return b
}

func (c *CarBuilder) SetStructure() VehicleBuilder {
	c.v.Structure = "Car"
	return c
}

func (c *CarBuilder) Build() Vehicle {
	return c.v
}









